### DISCUSSION TITLE
How to Style Badge Corners Based on Position and Center Toolbar Alignment in GrapesJS

### DISCUSSION THREAD

**Discussion by fredrickreuben** [GrapesJS/grapesjs#5962] [Q&A]
"First, I would like to style my badge so that when it is positioned at the top, the top-left and top-right corners are rounded. When the badge is positioned at the bottom of the component, the bottom-left and bottom-right corners should be rounded. Additionally, I want to center the toolbar so that it is aligned in the middle. How can I achieve this?"

**Comment by glaprida** [comment]
To center the toolbar you can add these styles

```
.gjs-toolbar{
    left: 50%; // you might need to add !important
    translate: -50% 0;
}
```

the rounding of the badge seems more complicated... css-wise its just a `.gjs-badge{ border-radius: [your values] }`, but to style it differently when its shown on top than when its shown at the bottom, that's the hard part, I guess. (I'm not sure how hard, but it's not just CSS... there's definitely JS involved)

**Reply by fredrickreuben** [reply]
thanks for your answer, !important does not work because of this inline rule element.style {
    pointer-events: all;
    top: 100px;
    left: 936px;
}

**Reply by glaprida** [reply]
`!important` should work (it works for me on dev tools)

<img width="352" alt="Screenshot 2024-06-23 at 21 13 58" src="https://github.com/GrapesJS/grapesjs/assets/10442382/ebfa719b-23d7-44a5-8a00-95f1efc7fc3c">

 
in fact, `!important` is the only way to override inline styles (unless they are marked as !important too, then you just cannot override them) https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity
 
<img width="841" alt="Screenshot 2024-06-23 at 21 17 49" src="https://github.com/GrapesJS/grapesjs/assets/10442382/01839967-a1fa-428a-b2c1-40401d5538cc">

**Reply by fredrickreuben** [reply]
I've managed to center the toolbar, but for the badge, I want to adjust the badge container. I'm considering how to add an attribute or class to the body element within the gjs-tools class, indicating whether the position is currently at the top or bottom. This way, I can apply corresponding styles. Any suggestions on how to achieve this?

**Reply by fredrickreuben** [reply]
Also, i would like the badge of the selected model to be hidden.

**Comment by fredrickreuben** [comment]
const onToolbarStyleChange = useCallback(() => {
        const el = document.getElementById('gjs-tools');

        if (!el) return;

        const handleStyleChanges = (mutationsList: any) => {
            for (let mutation of mutationsList) {
                if (
                    mutation.type === 'attributes' &&
                    mutation.attributeName === 'style'
                ) {
                    if (
                        mutation.type === 'attributes' &&
                        mutation.attributeName === 'style'
                    ) {
                        const target = mutation.target as HTMLElement;
                        const badge = target.querySelector('.gjs-toolbar');
                        if (!badge) return;
                        const { style } = badge as HTMLElement;
                        const topValue = parseInt(style.top || '0', 10);
                        if (topValue < 0) {
                            badge.classList.add('gjs-toolbar-top');
                            badge.classList.remove('gjs-toolbar-bottom');
                        } else {
                            badge.classList.add('gjs-toolbar-bottom');
                            badge.classList.remove('gjs-toolbar-top');
                        }
                    }
                }
            }
        };

        const observer = new MutationObserver(handleStyleChanges);

        const config = {
            attributes: true,
            attributeFilter: ['style'],
        };

        observer.observe(el, config);

    }, []);

    const onBadgeStyleChange = useCallback(() => {
        const el = document.getElementById('gjs-cv-tools');
        const toolsOnly = el?.querySelectorAll('.gjs-tools:not(.gjs-tools-gl)');

        const handleStyleChanges = (mutationsList: any) => {
            for (let mutation of mutationsList) {
                if (
                    mutation.type === 'attributes' &&
                    mutation.attributeName === 'style'
                ) {
                    if (
                        mutation.type === 'attributes' &&
                        mutation.attributeName === 'style'
                    ) {
                        const target = mutation.target as HTMLElement;
                        const badge = target.querySelector('.gjs-badge');
                        if (!badge) return;
                        const { style } = badge as HTMLElement;
                        const topValue = parseInt(style.top || '0', 10);
                        if (topValue < 0) {
                            badge.classList.add('gjs-top-rounded');
                            badge.classList.remove('gjs-bottom-rounded');
                        } else {
                            badge.classList.add('gjs-bottom-rounded');
                            badge.classList.remove('gjs-top-rounded');
                        }
                    }
                }
            }
        };

        const observer = new MutationObserver(handleStyleChanges);

        const config = {
            attributes: true,
            attributeFilter: ['style'],
        };

        toolsOnly?.forEach((tool) => {
            observer.observe(tool, config);
        });
    }, []);

    useEffect(() => {
        if (ready) {
            onBadgeStyleChange();
            onToolbarStyleChange()
        }
    }, [ready]);

**Reply by fredrickreuben** [reply]
This what i came up with, please rate it for me?
